xarray/core/missing.py

```diff
@@ -0,0 +1,293 @@
+            self._right = yi[-1]
+        elif isinstance(fill_value, Iterable) and len(fill_value) == 2:
+            self._left = fill_value[0]
+            self._right = fill_value[1]
+        elif is_scalar(fill_value):
+            self._left = fill_value
+            self._right = fill_value
+        else:
+            raise ValueError('%s is not a valid fill_value' % fill_value)
+
+    def __call__(self, x):
+        return self.f(x, self._xi, self._yi, left=self._left,
+                      right=self._right, **self.call_kwargs)
+
+
+class ScipyInterpolator(BaseInterpolator):
+    '''Interpolate a 1-D function using Scipy interp1d
+
+    See Also
+    --------
+    scipy.interpolate.interp1d
+    '''
+    def __init__(self, xi, yi, kind=None, fill_value=None, assume_sorted=True,
+                 copy=False, bounds_error=False, **kwargs):
+        from scipy.interpolate import interp1d
+
+        if kind is None:
+            raise ValueError('kind is a required argument')
+
+        if kind == 'polynomial':
+            kind = kwargs.pop('order', None)
+            if kind is None:
+                raise ValueError('order is required when method=polynomial')
+
+        self.kind = kind
+
+        self.cons_kwargs = kwargs
+        self.call_kwargs = {}
+
+        if fill_value is None and kind == 'linear':
+            fill_value = kwargs.pop('fill_value', (np.nan, yi[-1]))
+        elif fill_value is None:
+            fill_value = np.nan
+
+        self.f = interp1d(xi, yi, kind=self.kind, fill_value=fill_value,
+                          bounds_error=False, **self.cons_kwargs)
+
+
+class SplineInterpolator(BaseInterpolator):
+    '''One-dimensional smoothing spline fit to a given set of data points.
+
+    See Also
+    --------
+    scipy.interpolate.UnivariateSpline
+    '''
+    def __init__(self, xi, yi, kind=None, fill_value=None, order=3, **kwargs):
+        from scipy.interpolate import UnivariateSpline
+
+        if kind is None:
+            raise ValueError('kind is a required argument')
+
+        self.kind = kind
+        self.cons_kwargs = kwargs
+        self.call_kwargs['nu'] = kwargs.pop('nu', 0)
+        self.call_kwargs['ext'] = kwargs.pop('ext', None)
+
+        if fill_value is not None:
+            raise ValueError('SplineInterpolator does not support fill_value')
+
+        self.f = UnivariateSpline(xi, yi, k=order, **self.cons_kwargs)
+
+
+def get_clean_interp_index(arr, dim, use_coordinate=True, **kwargs):
+    '''get index to use for x values in interpolation.
+
+    If use_coordinate is True, the coordinate that shares the name of the
+    dimension along which interpolation is being performed will be used as the
+    x values.
+
+    If use_coordinate is False, the x values are set as an equally spaced
+    sequence.
+    '''
+
+    if use_coordinate:
+        if use_coordinate is True:
+            index = arr.get_index(dim)
+        else:
+            index = arr.coords[use_coordinate]
```

Comment:
Do we need to check `arr.coords[use_coordinate]` is certainly 1-dimensional?

Reply:
Maybe we can also raise an appropriate error if coordinate is `MultiIndex`

Reply:
@fujiisoup - do we have a handy way to tell if an coordinate is a xarray DataArray/Multindex? Its not a subset of `pd.Multindex` but I need something like:

```Python
isinstance(index, (pd.MultiIndex, xr.MultiIndexCoordinate))
```

Reply:
Probably the cleanest and most comprehensive thing to do is to try to always cast the interpolation index to `float64`, while catching any error messages (to replace them with something more informative). When a `MultiIndex` is cast to numpy array it ends up with object dtype.